examples/headsail-bsp/src/uart.rs

```diff
@@ -12,3 +17,26 @@ pub fn putc(c: u8) {
     // Safety: we don't know if u8 writes work for all target architectures
     unsafe { write_u8(UART0_ADDR, c) };
 }
+
+#[inline]
+pub fn getc() -> Option<u8> {
+    let data_ready = unsafe { read_u8(UART0_ADDR + UART_DATA_READY_OFFSET) };
+    if data_ready & 1 == 0 {
+        None
+    } else {
+        let byte = unsafe { read_u8(UART0_ADDR) };
+        Some(byte)
+    }
```

Comment:
I wonder if this should be more along the lines of 

```rust
// Wait for data to become ready
while data_ready & 1 == 0 {}

// SAFETY: UART0_ADDR is 4-byte aligned
unsafe {read_u8(UART0_ADDR)};
```

This will also simplify caller at `uart_read_to_heap`.